packages/flutter/lib/src/widgets/selection_container.dart

```diff
@@ -0,0 +1,207 @@
+
+import 'framework.dart';
+
+/// A container that handles selection events for the [Selectable]s in
+/// the subtree.
+///
+/// This widget is useful when one wants to customize selection behaviors for
+/// a group of [Selectable]s
+///
+/// The render object of this container is a single selectable and will register
+/// itself to the [registrar]. The containers handle the [SelectionEvent]s from
+/// the [registrar] and delegate the events to the [delegate].
+///
+/// This widget uses [SelectionRegistrarScope] to host the [delegate] as the
+/// [SelectionRegistrar] for the subtree to collect the [Selectable]s, and
+/// [SelectionEvent]s received by this container are sent to the [delegate] using
+/// the [SelectionHandler] API of the delegate.
+///
+/// To use this widget, set the [registrar] this container received the
+/// [SelectionEvent]s from. Use [SelectionRegistrarScope.maybeOf] to retrieve
+/// the registrar that is relevant to the current context, and then create a
+/// subclass of [SelectionContainerDelegate] to decide how the selections are
+/// handled.
+///
+/// {@tool dartpad}
+/// This sample demonstrates how to create a select-all-or-none container
+///
+/// ** See code in examples/api/lib/material/selection_area/custom_container.dart **
+/// {@end-tool}
+///
+/// See also:
+///  * [SelectableRegion], which provides an overview of the selection system.
+///  * [SelectionRegistrarScope.disabled], which disable selection for a
+///    subtree.
+class SelectionContainer extends StatefulWidget {
+  /// Creates a selection container to collect the [Selectable]s in the subtree.
+  ///
+  /// The [registrar] and [delegate] must not be null.
+  const SelectionContainer({
+    super.key,
+    required this.registrar,
+    required this.delegate,
+    required this.child
+  }) : assert(registrar != null && delegate != null);
+
+  /// The [SelectionRegistrar] this container is registered to.
+  final SelectionRegistrar registrar;
+
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  /// The delegate for [SelectionEvent]s sent to this selection container.
+  ///
+  /// The [Selectable]s in the subtree is added or removed from this delegate
+  /// uses [SelectionRegistrar] API.
+  ///
+  /// This delegate is responsible for updating the selections for the selectables
+  /// under this widget.
+  final SelectionContainerDelegate delegate;
+
+  @override
+  State<SelectionContainer> createState() => _SelectionContainerState();
+}
+
+class _SelectionContainerState extends State<SelectionContainer> with Selectable, SelectionRegistrant {
+  final Set<Selectable> _registeredSelectables = <Selectable>{};
+  late final ValueNotifier<SelectionGeometry> _notifier;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.delegate._selectionContainerContext = context;
+    widget.delegate.addListener(_onSelectionGeometryChange);
+    _notifier = ValueNotifier<SelectionGeometry>(widget.delegate.value);
```

Comment:
The delegate implements SelectionHandler, which itself implements ValueListenable. Why do we need to wrap this in another ValueNotifier? Can you just add/remove listeners to the delegate directly and access its value?

Reply:
right, not sure what I was thinking.